components/raftstore/src/store/worker/read.rs

```diff
@@ -207,13 +213,66 @@ where
     }
 }
 
-/// #[RaftstoreCommon]: LocalReadContext combines some LocalReader's fields for temporary usage.
 pub struct LocalReadContext<'a, E>
 where
     E: KvEngine,
 {
-    read_id: &'a mut ThreadReadId,
-    snap_cache: &'a mut Box<Option<Arc<E::Snapshot>>>,
+    read_id: Option<ThreadReadId>,
+    snap_cache: &'a mut SnapCache<E>,
+}
+
+impl<'a, E> LocalReadContext<'a, E>
+where
+    E: KvEngine,
+{
+    fn new(snap_cache: &'a mut SnapCache<E>, read_id: Option<ThreadReadId>) -> Self {
+        Self {
+            snap_cache,
+            read_id,
+        }
+    }
+
+    // Update the snapshot in the `snap_cache` if the read_id is None or does not
+    // match.
+    // When the read_id is None, it means the `snap_cache` has been cleared
+    // before and the `cached_read_id` of it is None because only a consecutive
+    // requests will have the same cache and the cache will be cleared after the
+    // last request of the batch.
+    fn maybe_update_snapshot(&mut self, engine: &E, delegate_last_valid_ts: Timespec) -> bool {
+        if let Some(read_id) = self.read_id.as_ref() {
+            if self.snap_cache.cached_read_id.is_some()
+                && *self.snap_cache.cached_read_id.as_ref().unwrap() == *read_id
+                && self.snap_cache.cached_snapshot_ts >= delegate_last_valid_ts
```

Comment:
Should compare read_id's create time instead of snapshot time. For example, if the snapshot is taken before the peer is created, the check can pass, but no data can be read.